sale_order_product_assortment/models/sale_order.py

```diff
@@ -19,61 +19,68 @@ class SaleOrder(models.Model):
                     )
-                    for domain in filters_partner_domain.mapped("domain")
                 )
-                whitelist_product_domain = expression.OR(
-                    [
-                        eval_product_domain,
+                if filters_partner_domain:
+                    # use OR to combine allowed products
+                    # and negate the blacklisted ones with AND
+                    eval_product_domain = expression.OR(
+                        safe_eval(
+                            domain,
+                            {
+                                "datetime": datetime,
+                                "context_today": datetime.datetime.now,
+                            },
+                        )
+                        for domain in filters_partner_domain.mapped("domain")
+                    )
+                    whitelist_product_domain = expression.OR(
                         [
-                            (
-                                "id",
-                                "in",
-                                filters_partner_domain.mapped(
-                                    "whitelist_product_ids.id"
-                                ),
-                            )
-                        ],
+                            eval_product_domain,
+                            [
+                                (
+                                    "id",
+                                    "in",
+                                    filters_partner_domain.mapped(
+                                        "whitelist_product_ids.id"
+                                    ),
+                                )
+                            ],
+                        ]
+                    )
+                    blacklist_product_domain = [
+                        (
+                            "id",
+                            "not in",
+                            filters_partner_domain.mapped("blacklist_product_ids.id"),
+                        )
                     ]
-                )
-                blacklist_product_domain = [
-                    (
-                        "id",
-                        "not in",
-                        filters_partner_domain.mapped("blacklist_product_ids.id"),
+                    product_domain = expression.AND(
+                        [whitelist_product_domain, blacklist_product_domain]
                     )
-                ]
-                product_domain = expression.AND(
-                    [whitelist_product_domain, blacklist_product_domain]
-                )
-            if product_domain:
-                self.allowed_product_ids = self.env["product.product"].search(
-                    product_domain
-                )
-            if self.allowed_product_ids:
-                self.has_allowed_products = True
+                    products = self.env["product.product"].search(product_domain)
```

Comment:
I would say, doing a search once before the loop should be the aim to get to.

IMHO, it is possible to do it, then filtering the products

Reply:
The vast, vast majority of the times (every time?) this method is executed with only one SO in `self`.

Either way, how would you suggest to go about it? We have over 50k products, and filtering them in python afterwards seems like it would slow things down significantly.